.cloud-build/execute_changed_notebooks_cli.py

```diff
@@ -129,26 +139,65 @@ def str2bool(v):
     base_branch=args.base_branch,
 )
 
+
+def _load_results() -> list:
+    '''
+    Load accumulated notebook test results
+    '''
+    print("Loading existing accumulative results ...")
+    rows = []
+    try:
+        df = pd.read_csv(os.path.join(f"gs://{args.artifacts_bucket}", args.test_results))
+        df = df.reset_index()
+
+        for index, row in df.iterrows():
+            rows.append([row["notebook"], row["duration"], row["passed"], row["failed"]])
+        print(rows)
+    except Exception as e:
+        print(e)
+
+    return rows
+
+def _select_notebook(changed_notebook: str) -> float:
+    '''
+    Algorithm to randomly select a notebook, but weight the liklihood based on past failures
+    '''
+    passed = 1
+    failed = 0
+    for notebook_result in notebook_results:
+        if notebook_result[0] == changed_notebook:
+            passed = notebook_result[2]
+            failed = notebook_result[3]
+            break
+
+    if not failed:
+        return random.randint(1, 100) * passed
+    else:
+        return randinit(1, 100) / failed
+
 if args.test_percent == 100:
     notebooks = changed_notebooks
 else:
-    notebooks = [changed_notebook for changed_notebook in changed_notebooks if random.randint(1, 100) < args.test_percent]
+    notebook_results = _load_results()
+
+    notebooks = [changed_notebook for changed_notebook in changed_notebooks if _select_notebook(changed_notebook) < args.test_percent]
```

Comment:
Small nit: this is a nested loop, so the time-complexity of this can be optimized by creating a counter dict over past results first and then iterating over notebooks.